storage/index/types.go

```diff
@@ -183,41 +178,111 @@ type BlockTickResult struct {
 	NumDocs     int64
 }
 
+// WriteBatchEntry captures a document to index, and the lifecycle hooks to call thereafter.
+type WriteBatchEntry struct {
+	ID            ident.ID
+	Tags          ident.Tags
+	Timestamp     time.Time
+	OnIndexSeries OnIndexSeries
+}
+
+// WriteBatchEntriesFinalizer is a utility type to provide syntactic sugar to finalize references in the slice.
+type WriteBatchEntriesFinalizer []WriteBatchEntry
+
+// Finalize finalizes all the references in the provided slice.
+func (w WriteBatchEntriesFinalizer) Finalize() {
+	for _, entry := range w {
+		if entry.OnIndexSeries != nil {
```

Comment:
need this check cause we set entries to the empty value if they don't need to be indexed (in `index.go:InsertBatch`)